thorlcr/graph/thgraph.cpp

```diff
@@ -2771,10 +2761,60 @@ void CJobBase::startJob()
 
 void CJobBase::endJob()
 {
+    if (jobEnded)
+        return;
+
+    jobEnded = true;
     setPerformanceMonitorHook(nullptr);
     LOG(MCdebugProgress, thorJob, "Job ended : %s", graphName.get());
     clearKeyStoreCache(true);
     PrintMemoryStatusLog();
+
+    Owned<IMultiException> exceptions;
+    ForEachItemIn(c, jobChannels)
+    {
+        try
+        {
+            jobChannels.item(c).clean();
+        }
+        catch (IException *e)
+        {
+            if (!exceptions)
+                exceptions.setown(makeMultiException());
+            exceptions->append(*LINK(e));
+        }
+    }
+
```

Comment:
Should you check for exceptions here and if present throw ?

Reply:
no, want to clearup rest first.
exceptions are checked at end (line 2805) if there any it will throw then.